File: LeetCode/Python/src/ZigZag_Conversion.py
"""
@author - Anirudh Sharma
"""


def convert(s: str, numRows: int) -> str:
    # Base condition
    if s is None and numRows <= 0:
        return ""
    if numRows == 1:
        return s
    # Resultant string
    result = ""
    # Step size
    step = 2 * numRows - 2
    # Loop for each row
    for i in range(0, numRows):
        # Loop for each character in the row
        for j in range(i, len(s), step):
            result += s[j]
            if i != 0 and i != numRows - 1 and (j + step - 2 * i) < len(s):
                result += s[j + step - 2 * i]
    return result


if __name__ == '__main__':
    s = "PAYPALISHIRING"
    numRows = 3
    print(convert(s, numRows))

    numRows = 4
    print(convert(s, numRows))

    s = "ANIRUDHSHARMAISGREAT"
    numRows = 5
    print(convert(s, numRows))

File: LeetCode/JavaScript/src/zigzag_conversion.js
/**
 * @author Anirudh Sharma
 */
var convert = function (s, numRows) {
    // Base conditions
    if (s === null || numRows <= 0) {
        return "";
    }
    if (numRows === 1) {
        return s;
    }
    // Resultant String
    let result = ""
    // Step size
    const step = 2 * numRows - 2;
    // Loop for each row
    for (let i = 0; i < numRows; i++) {
        // Loop for each character in the row
        for (let j = i; j < s.length; j += step) {
            result += s[j];
            if (i != 0 && i != numRows - 1 && (j + step - 2 * i) < s.length) {
                result += s[j + step - 2 * i];
            }
        }
    }
    return result;
};

let s = "PAYPALISHIRING"
let numRows = 3;
console.log(convert(s, numRows));

numRows = 4;
console.log(convert(s, numRows));

s = "ANIRUDHSHARMAISGREAT";
numRows = 5;
console.log(convert(s, numRows));